File: tests/Billing.Tests/BasicChargeTests.cs
using Billing.Mediation.Context;
using Billing.Mediation.Model;
using Billing.Mediation.Rating;
using MediationModel;

namespace Billing.Tests;

/// <summary>The basic charge end to end: detect SG → resolve the rate-plan tuple → longest-prefix the
/// normalized number in the today RateCache → A2Z duration+amount. Proves the whole rating chain wires
/// together over a MediationContext.</summary>
public class BasicChargeTests
{
    // Plan 7 holds a per-minute rate for prefix 1712 (national-significant form of 8801712...).
    private static MediationContext Mediation() => new()
    {
        RateCache = RateCache.Build(new Dictionary<int, IReadOnlyDictionary<string, Rate>>
        {
            [7] = new Dictionary<string, Rate>
            {
                ["1712"] = new() { Prefix = "1712", RateAmount = 1.0m, IdRatePlan = 7 },
            },
        }, new DateOnly(2026, 6, 19)),
        RatePlanResolver = RatePlanResolver.Build(new[]
        {
            new RatePlanAssignmentTuple
            {
                Id = 1, IdService = 10, AssignDirection = (int)AssignmentDirection.Customer,
                IdPartner = 5, IdRatePlan = 7,
            },
        }),
    };

    private static IReadOnlyDictionary<int, Partner> Partners(params Partner[] ps) => ps.ToDictionary(p => p.IdPartner);

    [Fact]
    public void Charges_a_detected_call_end_to_end()
    {
        var thisCdr = new cdr { InPartnerId = 5, TerminatingCalledNumber = "8801712345678", DurationSec = 60m };
        var result = BasicCharge.Default().Compute(
            thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner { IdPartner = 5, PartnerType = 3 }));

        Assert.NotNull(result);
        Assert.Equal(10, result.Value.ServiceGroupId);     // retail → SG10
        Assert.Equal(7, result.Value.IdRatePlan);
        Assert.Equal("1712", result.Value.MatchedPrefix);
        Assert.Equal(60m, result.Value.BilledDurationSec);
        Assert.Equal(1.0m, result.Value.Amount);           // 60s per-minute @ 1.0
    }

    [Fact]
    public void Half_minute_is_half_the_per_minute_rate()
    {
        var thisCdr = new cdr { InPartnerId = 5, TerminatingCalledNumber = "8801712345678", DurationSec = 30m };
        var result = BasicCharge.Default().Compute(
            thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner { IdPartner = 5, PartnerType = 3 }));

        Assert.Equal(30m, result!.Value.BilledDurationSec);
        Assert.Equal(0.5m, result.Value.Amount);
    }

    [Fact]
    public void No_charge_when_service_group_not_detected()
    {
        var thisCdr = new cdr { InPartnerId = 5, TerminatingCalledNumber = "8801712345678", DurationSec = 60m };
        // partnerType 1 → neither SG10 nor SG11 claims
        var result = BasicCharge.Default().Compute(
            thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner { IdPartner = 5, PartnerType = 1 }));
        Assert.Null(result);
    }

    [Fact]
    public void No_charge_when_no_rate_plan_assigned()
    {
        var thisCdr = new cdr { InPartnerId = 6, TerminatingCalledNumber = "8801712345678", DurationSec = 60m };
        // SG10 detected (retail) but no tuple for partner 6 → no plan
        var result = BasicCharge.Default().Compute(
            thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner { IdPartner = 6, PartnerType = 3 }));
        Assert.Null(result);
    }

    [Fact]
    public void No_charge_when_number_matches_no_rate_prefix()
    {
        var thisCdr = new cdr { InPartnerId = 5, TerminatingCalledNumber = "8809999999", DurationSec = 60m };
        // resolves plan 7, but 9999999 has no prefix in the plan
        var result = BasicCharge.Default().Compute(
            thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner { IdPartner = 5, PartnerType = 3 }));
        Assert.Null(result);
    }
}

File: java/src/test/java/com/telcobright/billing/mediation/rating/BasicChargeTests.java
package com.telcobright.billing.mediation.rating;

import com.telcobright.billing.mediation.context.MediationContext;
import com.telcobright.billing.mediation.engine.models.cdr;
import com.telcobright.billing.mediation.model.AssignmentDirection;
import com.telcobright.billing.mediation.model.Partner;
import com.telcobright.billing.testsupport.TestData;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

/**
 * The customer-leg charge end to end over the legacy entities: detect SG -&gt; resolve the tuple -&gt;
 * PrefixMatcher -&gt; the SG's service family -&gt; acc_chargeable. Proves the whole rating chain wires together
 * over a MediationContext.
 */
class BasicChargeTests {

    // SG10 customer tuple for partner 5: per-minute 1.0 for prefix 1712, rate plan 7.
    private static MediationContext Mediation() {
        var f = TestData.fixture();
        f.tup(10, AssignmentDirection.Customer.value, 5, null, 0, TestData.Ra(8801712, "1.0").idRatePlan(7));
        return f.mediation();
    }

    private static Map<Integer, Partner> Partners(Partner... ps) {
        Map<Integer, Partner> m = new HashMap<>();
        for (Partner p : ps) m.put(p.IdPartner(), p);
        return m;
    }

    @Test
    void Charges_a_detected_call_end_to_end() {
        var thisCdr = new cdr();
        thisCdr.InPartnerId = 5;
        thisCdr.TerminatingCalledNumber = "8801712345678"; thisCdr.OriginatingCalledNumber = "8801712345678";
        thisCdr.DurationSec = new BigDecimal("60");
        var chargeable = BasicCharge.Default().Compute(
                thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner(5, null, 3)));

        assertNotNull(chargeable);
        assertEquals(10, chargeable.servicegroup);     // retail -> SG10
        assertEquals(10, chargeable.servicefamily);    // SF10
        assertEquals("8801712", chargeable.Prefix);
        assertEquals(0, new BigDecimal("60").compareTo(chargeable.Quantity));
        assertEquals(0, new BigDecimal("1.0").compareTo(chargeable.BilledAmount));
    }

    @Test
    void Half_minute_is_half_the_per_minute_rate() {
        var thisCdr = new cdr();
        thisCdr.InPartnerId = 5;
        thisCdr.TerminatingCalledNumber = "8801712345678"; thisCdr.OriginatingCalledNumber = "8801712345678";
        thisCdr.DurationSec = new BigDecimal("30");
        var chargeable = BasicCharge.Default().Compute(
                thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner(5, null, 3)));

        assertEquals(0, new BigDecimal("30").compareTo(chargeable.Quantity));
        assertEquals(0, new BigDecimal("0.5").compareTo(chargeable.BilledAmount));
    }

    @Test
    void No_charge_when_service_group_not_detected() {
        var thisCdr = new cdr();
        thisCdr.InPartnerId = 5;
        thisCdr.TerminatingCalledNumber = "8801712345678"; thisCdr.OriginatingCalledNumber = "8801712345678";
        thisCdr.DurationSec = new BigDecimal("60");
        var chargeable = BasicCharge.Default().Compute(
                thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner(5, null, 1)));
        assertNull(chargeable);
    }

    @Test
    void No_charge_when_no_rate_plan_assigned() {
        var thisCdr = new cdr();
        thisCdr.InPartnerId = 6;
        thisCdr.TerminatingCalledNumber = "8801712345678"; thisCdr.OriginatingCalledNumber = "8801712345678";
        thisCdr.DurationSec = new BigDecimal("60");
        // SG10 detected (retail) but no tuple for partner 6
        var chargeable = BasicCharge.Default().Compute(
                thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner(6, null, 3)));
        assertNull(chargeable);
    }

    @Test
    void No_charge_when_number_matches_no_rate_prefix() {
        var thisCdr = new cdr();
        thisCdr.InPartnerId = 5;
        thisCdr.TerminatingCalledNumber = "8809999999"; thisCdr.OriginatingCalledNumber = "8809999999";
        thisCdr.DurationSec = new BigDecimal("60");
        var chargeable = BasicCharge.Default().Compute(
                thisCdr, AssignmentDirection.Customer, Mediation(), Partners(new Partner(5, null, 3)));
        assertNull(chargeable);
    }
}
